### PR TITLE
Added HTTPC POST request example.

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by WinterMute** [comment]
You have  a note in here that says `Enable Keep-Alive connections (on by default, pending ctrulib merge)`. I don't see a pending libctru PR, do you have something to submit there?

**Comment by ghost** [comment]
@WinterMute I have taken the already existing http example as my base hete. I'mma look into that PR.

**Comment by ghost** [comment]
Here is [that PR](https://github.com/smealum/ctrulib/pull/301). Let me edit the code a bit there.

**Comment by fincs** [comment]
gfxFlushBuffers isn't required if you print a newline.

**Comment by WinterMute** [comment]
I see, this makes a bit more sense now I look at the http example - I've fixed the overuse of gfxFlushBuffers there now.

**Comment by ghost** [comment]
Also removed it.

File: network/http_post/source/main.c [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	char *newurl=NULL;
+	u32 statuscode=0;
+	u32 contentsize=0, readsize=0, size=0;
+	u8 *buf, *lastbuf;
+
+	printf("POSTing %s\n", url);
+
+	do {
+		ret = httpcOpenContext(&context, HTTPC_METHOD_POST, url, 0);
+		printf("return from httpcOpenContext: %"PRIx32"\n",ret);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: need spaces around string literal here. i.e.  %" PRIx32 "\n"

File: network/http_post/source/main.c [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+		// This disables SSL cert verification, so https:// will be usable
+		ret = httpcSetSSLOpt(&context, SSLCOPT_DisableVerify);
+		printf("return from httpcSetSSLOpt: %"PRIx32"\n",ret);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+		// Enable Keep-Alive connections (on by default, pending ctrulib merge)
+		ret = httpcSetKeepAlive(&context, HTTPC_KEEPALIVE_ENABLED);
+		printf("return from httpcSetKeepAlive: %"PRIx32"\n",ret);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+		// Set a User-Agent header so websites can identify your application
+		ret = httpcAddRequestHeaderField(&context, "User-Agent", "httpc-example/1.0.0");
+		printf("return from httpcAddRequestHeaderField: %"PRIx32"\n",ret);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+		// Set a Content-Type header so websites can identify the format of our raw body data.
+		// If you want to send form data in your request, use:
+		//ret = httpcAddRequestHeaderField(&context, "Content-Type", "multipart/form-data");
+		// If you want to send raw JSON data in your request, use:
+		ret = httpcAddRequestHeaderField(&context, "Content-Type", "application/json");
+		printf("return from httpcAddRequestHeaderField: %"PRIx32"\n",ret);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+		// Post specified data.
+		// If you want to add a form field to your request, use:
+		//ret = httpcAddPostDataAscii(&context, "data", value);
+		// If you want to add a form field containing binary data to your request, use:
+		//ret = httpcAddPostDataBinary(&context, "field name", yourBinaryData, length);
+		// If you want to add raw data to your request, use:
+		ret = httpcAddPostDataRaw(&context, (u32*)data, strlen(data));
+		printf("return from httpcAddPostDataRaw: %"PRIx32"\n",ret);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+			}
+			ret = httpcGetResponseHeader(&context, "Location", newurl, 0x1000);
+			url = newurl; // Change pointer to the url that we just learned
+			printf("redirecting to url: %s\n",url);
+			httpcCloseContext(&context); // Close this context before we try the next
+		}
+	} while ((statuscode >= 301 && statuscode <= 303) || (statuscode >= 307 && statuscode <= 308));
+
+	if(statuscode!=200){
+		printf("URL returned status: %"PRIx32"\n", statuscode);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+int main()
+{
+	Result ret=0;
+	gfxInitDefault();
+	httpcInit(4 * 1024 * 1024); // Buffer size when POST/PUT.
+
+	consoleInit(GFX_BOTTOM,NULL);
+
+	ret=http_post("http://httpbin.org/post", "{\"foo\": \"bar\"}");
+	printf("return from http_post: %"PRIx32"\n",ret);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+	lastbuf = buf;
+	buf = realloc(buf, size);
+	if(buf==NULL){ // realloc() failed.
+		httpcCloseContext(&context);
+		free(lastbuf);
+		if(newurl!=NULL) free(newurl);
+		return -1;
+	}
+
+	printf("response size: %"PRIx32"\n",size);
```
Comment (WinterMute) [inline network/http_post/source/main.c]: spaces around literal here

File: network/http_post/source/main.c [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+	// Print result
+	printf((char*)buf);
+
+	gfxFlushBuffers();
+	gfxSwapBuffers();
+	gspWaitForVBlank();
```
Comment (WinterMute) [inline network/http_post/source/main.c]: this segment of code shoudn't be here. I think a printf(\n"); might be more appropriate just to make sure the complete contents of buf are printed.

File: network/http_post/source/main.c [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+{
+	Result ret=0;
+	gfxInitDefault();
+	httpcInit(4 * 1024 * 1024); // Buffer size when POST/PUT.
+
+	consoleInit(GFX_BOTTOM,NULL);
+
+	ret=http_post("http://httpbin.org/post", "{\"foo\": \"bar\"}");
+	printf("return from http_post: %"PRIx32"\n",ret);
+	gfxFlushBuffers();
```
Comment (WinterMute) [inline network/http_post/source/main.c]: this line isn't necessary

**Review by WinterMute — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Made requested changes.

File: network/http_post/source/main.c (hunk 7 above)
Comment (WinterMute) [reply network/http_post/source/main.c]: Missed this one

**Comment by WinterMute** [comment]
@moriczgergo Not sure if commenting in the commit pinged you so commenting again.

You missed one of the literals, can still see it here https://github.com/devkitPro/3ds-examples/pull/20#pullrequestreview-53078826

File: network/http_post/source/main.c [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+	do {
+		ret = httpcOpenContext(&context, HTTPC_METHOD_POST, url, 0);
+		printf("return from httpcOpenContext: %" PRIx32 "\n",ret);
+
+		// This disables SSL cert verification, so https:// will be usable
+		ret = httpcSetSSLOpt(&context, SSLCOPT_DisableVerify);
+		printf("return from httpcSetSSLOpt: %" PRIx32 "\n",ret);
+
+		// Enable Keep-Alive connections (on by default, pending ctrulib merge)
```
Comment (WinterMute) [inline network/http_post/source/main.c]: And lose the end of this comment so it's just // Enable Keep-Alive connections

we're not pending a PR after all :p

**Comment by ghost** [comment]
Sorry about the delay, it's done now.

**Comment by WinterMute** [comment]
Still missing spaces round this literal https://github.com/devkitPro/3ds-examples/pull/20/files#diff-2b4a3f2c90bcf9c7a7a53b1c6023e087R79

I think that'll be everything

**Comment by ghost** [comment]
Whoops. Could've sworn I've fixed those.

**Review by WinterMute — APPROVED** [review]

**Pull request merged**